hosted backend api doc (#2115)
FusionAuth/fusionauth-site: site/docs/v1/tech/apis/hosted-backend.adoc

---
layout: doc
title: Hosted Backend APIs
description: APIs that you to perform OAuth 2 Authorization Code Flow login, logout, and refresh for applications
---

> == Overview

Reviewer: I tested this all with a simple HTML page with a handful of forms on it. It could be useful to include a plain HTML example of calling login, register, logout.

Author: I agree that would be useful but think it may be out of scope for the API documentation page. Would it be worth providing another fusionauth-example repo?

|500
|There was a FusionAuth internal error. A stack trace is provided and logged in the FusionAuth log files.
|===

== Register

This API will start an OAuth 2 authorization code flow by building a valid request and then redirecting the browser to our `/oauth2/register` endpoint. This endpoint is nearly identical to the <<login>> endpoint with the exception that the target is registration instead of login. If the user is not logged in the user will be presented with the registration page and prompted for credentials before being redirected back to the <<callback>> endpoint.

To use this API correctly the browser should be redirected to this route and it will respond with a redirect status code if successful.

=== Request

[.api-authentication]
link:/docs/v1/tech/apis/authentication#no-authentication-required[icon:unlock[type=fas]] Start the registration flow

[.endpoint]
.URI
--
[method]#GET# [uri]#/app/register/{clientId}``?redirect_uri=\{redirectUri\}&state=\{state\}&scope=\{scope\}``#
--

==== Request Parameters

[.api]
[field]#clientId# [type]#[UUID]# [required]#Required#::
The top-level client Id for your application.

> [field]#redirect_uri# [type]#[String]# [optional]#Optional#::

Reviewer: Let the reader know that if a value is provided, it needs to be in the list of authorized URLs for the application. What happens if this isn't provided, and there are no authorized URLs defined in the app?

Author: I added a link in the prereqs section about the application configuration which covers this in more detail.

+
Available scopes:
+
* `openid` - This scope is used to request the `app.idt` Id token cookie be returned in the response
* `offine_access` - This scope is used to request the `app.rt` refresh token cookie be returned in the response

.Example Request URL
[source,url]
----
https://auth.yourdomain.io/app/register/297ca84b-69a9-4508-8649-97644e1d0b3d?redirect_uri=https%3A%2F%2Fapp.yourdomain.io%2Fcallback&state=yourStateData&scope=offline_access
----

=== Response

Successful invocations of this route will return a  `302` redirect to `/oauth2/register`. Other status codes indicate an error.

[cols="1a,9a"]
.Response Codes
|===
|Code |Description

|200
|This is likely an error case. The route will serve up an error page with html and details on what went wrong.

|302
|A successful request will redirect the user to `/oauth2/register` to register.

|403
|A forbidden response typically means that the Origin of this request did not pass the FusionAuth CORS filter. You likely need to add your app to your link:/docs/v1/tech/reference/cors[CORS Configuration].

|500
|There was a FusionAuth internal error. A stack trace is provided and logged in the FusionAuth log files.
|===

> == Callback

Author: I did think about this, but my rationale for including is that the devs will set it when building and testing and it mirrors the API we instruct people to build in our example applications and elsewhere. Since we don't have another spot to document it right now I would like to leave it fully described here, acknowledging that it may be slightly confusing to some.

These endpoints will set the following cookies which are `Secure` and have a `SameSite` value of `Lax`. This follows our link:/learn/expert-advice/oauth/oauth-token-storage#client-side-storage[expert advice on client-side storage].

.Cookies Set By the Hosted Backend
[cols="1,1,3"]
|===
| Name | HttpOnly | Description

| app.at
| true
| The access token for the configured application. This is a JWT and can be presented to your APIs to access data and functionality.

| app.rt
| true
| The refresh token for the configured application. Only present if the `offline_access` scope is requested. This can be presented to FusionAuth to retrieve a new access token.

| app.idt
| false
| The Id token for the user for the configured application. Only present if the `openid` scope is requested. This is a JWT and can be accessed by JavaScript to display user account information.

| app.at_exp
| false
| The unix epoch timestamp indicating when the access token will expire. This can be checked by JavaScript to determine when a refresh token should be used to get a new access token.

|===

> FusionAuth will set the domain on these cookies to `.example.com` where `example` is the domain name that FusionAuth is serving from either from the domain or any subdomain, `com` is the top-level domain, and the `.` allows the cookie to match the domain and all subdomains. If the host is a simple host name or IP address FusionAuth will set the domain to that (i.e. `localhost` or `127.0.0.1`).

Reviewer 2: ```suggestion
FusionAuth will set the domain on these cookies to `.example.com` where `example` is the domain name that FusionAuth is serving from either from the domain or any subdomain, `com` is the top-level domain, and the `.` allows the cookie to match the domain and all subdomains. If the host is a simple host name or IP address FusionAuth will set the domain to that (i.e. `localhost` or `127.0.0.1`). If FusionAuth is on a nested domain, then it will set cookies on the broadest domain that is not a top-level domain.
```

What this means is that FusionAuth needs to be hosted on the same domain or a subdomain or sibling domain of the application that you intend to use with these endpoints.

> For example if your app is on `app.example.com` and FusionAuth is on `auth.example.com` the cookies would be usable by your application.

Author: `.example.com`. What do I need to say to make this more clear?

> For example if your app is on `app.example.com` and FusionAuth is on `auth.example.com` the cookies would be usable by your application.

Reviewer 2: ```suggestion
For example if your app is on `app.example.com` and FusionAuth is on `auth.example.com` the cookies would be usable by your application. If FusionAuth is on `auth.department.division.example.com` and the app lives on `app.otherdepartment.otherdivision.example.com`, the cookies would still be usable, since the cookies are set on the `example.com` domain.
```

> For example if your app is on `app.example.com` and FusionAuth is on `auth.example.com` the cookies would be usable by your application.

Reviewer 2: Added two suggestions that make it clearer to me.

The URL encoded URL that the browser will be redirected to at the end of the login flow. If provided, this URL must be included in the list of [field]#Authorized Redirect URLs# for your application. If not provided, the default will be the first value in the [field]#Authorized Redirect URLs# list defined for your application. This parameter is validated the same as if it were being passed to `/oauth/authorize`, however when using this endpoint FusionAuth will pass <<Callback>> as the [field]#redirect_uri# to `/oauth2/authorize` as that route will handle the token exchange.

[field]#state# [type]#[String]# [optional]#Optional#::
The value of this parameter will be echoed back in the [field]#state# parameter of the redirect URL at the end of the login flow.

[field]#scope# [type]#[String]# [optional]#Optional# [default]#defaults to `openid offline_access`#::
The OAuth2 [field]#scope# parameter to be passed to the `/oauth2/authorize` endpoint. The format is a URL encoded, space-separated list of scopes (i.e `openid+offline_access` or `openid%20offline_access`).
+
Available scopes:
+
* `openid` - This scope is used to request the `app.idt` Id token cookie be returned in the response

> * `offine_access` - This scope is used to request the `app.rt` refresh token cookie be returned in the response

Reviewer 3: There is a typo in `offline_access`.

The Id of the Tenant that is associated with the Application.

include::docs/v1/tech/shared/_oauth-authorize-redirect-parameters.adoc[]

.Example Request URL
[source,url]
----
https://auth.example.com/app/callback?code=wJfjafZLvo_KH5-D4r-3YwMmStN3yHoZDGmBivjioz0&locale=en&state=eyJjIjoiODVhMDM4NjctZGNjZi00ODgyLWFkZGUtMWE3&userState=Authenticated&client_id=297ca84b-69a9-4508-8649-97644e1d0b3d&tenantId=e707be45-afa8-4881-9efb-4be7288395d2
----

=== Response

A successful response should set cookies and return a `302` redirect to the `redirect_uri` specified in the initial request. Other status codes indicate an error.

[cols="1a,9a"]
.Response Codes
|===
|Code |Description

|200
|There was an error. The route will serve up an error page with html and details on what went wrong.

|302
|A successful request will redirect the user to `redirect_uri` specified in the request or the default [field]#Authorized Redirect URL# configured for the Application.

|500
|There was a FusionAuth internal error. A stack trace is provided and logged in the FusionAuth log files.
|===

==== Response Cookies

include::docs/v1/tech/shared/_hosted-backend-response-cookies.adoc[]

== Refresh

> This endpoint will extract the `app.rt` cookie if present and use it to make a `refresh_token` request from `/oauth/token`. If successful a new set of cookies will be set on the response that will continue to allow access to the application. You can call this any time or you can review the value of `app.at_exp` and call it when the access token is about to expire.

Reviewer 2: ```suggestion
This endpoint will extract the `app.rt` cookie if present and use it to make a `refresh_token` request from `/oauth/token`. The configuration rules for your Application configuration apply; ensure that [field]Refresh token# grant is enabled. If successful a new set of cookies will be set on the response that will continue to allow access to the application. You can call this any time or you can review the value of `app.at_exp` and call it when the access token is about to expire.
```

with PKCE. We have in the past shown you how to link:/blog/2021/11/11/how-to-authenticate-your-react-app#create-the-express-server[create these endpoints yourself]
but this solution allows you to get going with your app without writing any backend code. You just need FusionAuth!

* <<Prerequisites>>
* <<Login>>
* <<Register>>
* <<Callback>>
* <<Refresh>>
* <<Me>>
* <<Logout>>

== Prerequisites

Be sure to review the link:/docs/v1/tech/core-concepts/applications#oauth[Applications] section of the FusionAuth user guide to ensure proper configuration before using the hosted endpoints.

These endpoints will set the following cookies which are `Secure` and have a `SameSite` value of `Lax`. This follows our link:/learn/expert-advice/oauth/oauth-token-storage#client-side-storage[expert advice on client-side storage].
|===
| Name | HttpOnly | Description

| app.at
| true
| The access token (JWT) for the configured application

| app.rt
| true
| The refresh token for the configured applicaiton

| app.idt
| false
| The Id token (JWT) for the user for the configured application

| app.at_exp
| false
| The unix timestamp in seconds indicating when the access token will expire

|===

> FusionAuth will set the domain on these cookies to `.yourdomain.io` where `yourdomain` is the domain that FusionAuth is serving from, `io` is the top-level domain, and the `.` allows the cookie to match the domain and all subdomains. If the host is a simple host name or IP address FusionAuth will set the domain to that (i.e. `localhost` or `127.0.0.1`).

Author: attempted to add more clarity there

For example if your app is on `app.yourdomain.io` and FusionAuth is on `auth.yourdomain.io` the cookies would be usable by your application.

== Login

This API will start an OAuth 2 authorization code flow by building a valid request and then redirecting the browser to our `/oauth2/authorize` endpoint. If the user is not logged in the user will be presented with the login page and prompted for credentials before being redirected back to the <<callback>> endpoint.

To use this API correctly the browser should be redirected to this route and it will respond with a redirect status code if successful.

=== Request

[.api-authentication]
link:/docs/v1/tech/apis/authentication#no-authentication-required[icon:unlock[type=fas]] Start the login flow

[.endpoint]
.URI
--
[method]#GET# [uri]#/app/login/{clientId}``?redirect_uri=\{redirectUri\}&state=\{state\}&scope=\{scope\}``#
--

==== Request Parameters

[.api]
[field]#clientId# [type]#[UUID]# [required]#Required#::
The top-level client Id for your application.

[field]#redirect_uri# [type]#[String]# [optional]#Optional#::

> The URI that the browser will be redirected to at the end of the login flow. If provided this url must be defined in the list of Authorized Redirect URLs for your application. If not provided the default will be the first of the Authorized Redirect URLs defined for your application.

Author: added a description for why

+
Available scopes:
+
* `openid` - This scope is used to request the `app.idt` Id token cookie be returned in the response
* `offine_access` - This scope is used to request the `app.rt` refresh token cookie be returned in the response

.Example Request URL
[source,url]
----
https://auth.yourdomain.io/app/login/297ca84b-69a9-4508-8649-97644e1d0b3d?redirect_uri=https%3A%2F%2Fapp.yourdomain.io%2Fcallback&state=yourStateData&scope=offline_access
----

=== Response

Successful invocations of this route will return a  `302` redirect to `/oauth2/authorize`. Other status codes indicate an error.

[cols="1a,9a"]
.Response Codes
|===
|Code |Description

|200
|This is likely an error case. The route will serve up an error page with html and details on what went wrong.

|302
|A successful request will redirect the user to `/oauth2/authorize` to log in.

|403
|A forbidden response typically means that the Origin of this request did not pass the FusionAuth CORS filter. You likely need to add your app to your link:/docs/v1/tech/reference/cors[CORS Configuration].

|500
|There was a FusionAuth internal error. A stack trace is provided and logged in the FusionAuth log files.
|===

> 

Reviewer 2: We should document the response cookies like we do here: https://fusionauth.io/docs/v1/tech/apis/login#authenticate-a-user

+
Available scopes:
+
* `openid` - This scope is used to request the `app.idt` Id token cookie be returned in the response
* `offine_access` - This scope is used to request the `app.rt` refresh token cookie be returned in the response

.Example Request URL
[source,url]
----
https://auth.yourdomain.io/app/register/297ca84b-69a9-4508-8649-97644e1d0b3d?redirect_uri=https%3A%2F%2Fapp.yourdomain.io%2Fcallback&state=yourStateData&scope=offline_access
----

=== Response

Successful invocations of this route will return a  `302` redirect to `/oauth2/register`. Other status codes indicate an error.

[cols="1a,9a"]
.Response Codes
|===
|Code |Description

|200
|This is likely an error case. The route will serve up an error page with html and details on what went wrong.

|302
|A successful request will redirect the user to `/oauth2/register` to register.

|403
|A forbidden response typically means that the Origin of this request did not pass the FusionAuth CORS filter. You likely need to add your app to your link:/docs/v1/tech/reference/cors[CORS Configuration].

|500
|There was a FusionAuth internal error. A stack trace is provided and logged in the FusionAuth log files.
|===

> 

Reviewer 2: We should document the response cookies like we do here: https://fusionauth.io/docs/v1/tech/apis/login#authenticate-a-user

+
The possible values are:
+
* `Authenticated` - The user is successfully authenticated for the requested application.
* `AuthenticatedNotRegistered` - The user is successfully authenticated, but not registered for the requested application.

.Example Request URL
[source,url]
----
https://auth.yourdomain.io/app/callback?code=wJfjafZLvo_KH5-D4r-3YwMmStN3yHoZDGmBivjioz0&locale=en&state=eyJjIjoiODVhMDM4NjctZGNjZi00ODgyLWFkZGUtMWE3&userState=Authenticated&client_id=297ca84b-69a9-4508-8649-97644e1d0b3d&tenantId=e707be45-afa8-4881-9efb-4be7288395d2
----

=== Response

A successful response should set cookies and return a `302` redirect to the `redirect_uri` specified in the initial request. Other status codes indicate an error.

[cols="1a,9a"]
.Response Codes
|===
|Code |Description

|200
|This is likely an error case. The route will serve up an error page with html and details on what went wrong.

|302
|A successful request will redirect the user to `redirect_uri` specified in the request or the default Authorized Redirect URL configured for the application.

|500
|There was a FusionAuth internal error. A stack trace is provided and logged in the FusionAuth log files.
|===

> 

Reviewer 2: We should document the response cookies like we do here: https://fusionauth.io/docs/v1/tech/apis/login#authenticate-a-user (maybe)

== Refresh

> This endpoint will extract the `app.rt` cookie if present and use it to make a `refresh_token` request from `/oauth/token`. If successful a new set of cookies will be set on the response the will continue to allow access to the application.

Reviewer 2: ```suggestion
This endpoint will extract the `app.rt` cookie if present and use it to make a `refresh_token` request from `/oauth/token`. If successful a new set of cookies will be set on the response the will continue to allow access to the application. You can call this any time or you can review the value of `app.at_exp` and call it when the access token is about to expire.
```

This is an api that should be made from the client application and the browser should *NOT* be redirected to this endpoint.

=== Request

[.api-authentication]

> link:/docs/v1/tech/apis/authentication#jwt-authentication[icon:id-badge[type=fas]] Refresh the current session cookies.

Reviewer 2: It's not JWT authentication here, because it is the refresh token which allows access. I think you should add a new section to https://fusionauth.io/docs/v1/tech/apis/authentication or use `no authentication required`.

> link:/docs/v1/tech/apis/authentication#jwt-authentication[icon:id-badge[type=fas]] Refresh the current session cookies.

Reviewer 2: Unless I misunderstand how we are doing authentication here. Looked at the code for this action and don't see a JWT auth action.

> link:/docs/v1/tech/apis/authentication#jwt-authentication[icon:id-badge[type=fas]] Refresh the current session cookies.

Author: you are correct

[.endpoint]
.URI
--
[method]#POST# [uri]#/app/refresh/{clientId}#
--

==== Request Parameters

[.api]
[field]#client_id# [type]#[UUID]# [required]#Required#::
The top-level client Id for your application.

.Example Request URL
[source,url]
----
https://auth.yourdomain.io/app/refresh/297ca84b-69a9-4508-8649-97644e1d0b3d
----

=== Response

A successful response should set cookies and return a `200`.

[cols="1a,9a"]
.Response Codes
|===
|Code |Description

|200
|This is likely an error case. The route will serve up an error page with html and details on what went wrong.

|400

> |The request was not successful. The client may need to reauthorize.

Reviewer 2: ```suggestion
|The request was not successful. The client needs to reauthorize. Redirect the browser to the <<Login>> endpoint
```

|500
|There was a FusionAuth internal error. A stack trace is provided and logged in the FusionAuth log files.
|===

> 

Reviewer 2: We should document the response cookies like we do here: https://fusionauth.io/docs/v1/tech/apis/login#authenticate-a-user

== Me

This api is used to retrieve information about the currently logged in user. This call will take the `app.at` cookie value and use that to call the `/oauth2/userinfo` API.

This is an api that should be made from the client application and the browser should *NOT* be redirected to this endpoint.

=== Request

[.api-authentication]
link:/docs/v1/tech/apis/authentication#jwt-authentication[icon:id-badge[type=fas]] Get info about the user.

[.endpoint]
.URI
--
[method]#GET# [uri]#/app/me/#
--

.Example Request URL
[source,url]
----
https://auth.yourdomain.io/app/me
----

=== Response

A successful response should set cookies and return a `302` redirect to the `redirect_uri` specified in the initial request. Other status codes indicate an error.

[cols="1a,9a"]
.Response Codes
|===
|Code |Description

|200
|The request was successful. The response will contain a JSON body.

|401

> |The user is not authorized. May need to refresh or log in again.

Reviewer 2: ```suggestion
|The user is not authorized. Call <<Refresh>> or redirect the browser to the <<Login>> endpoint.
```

Author: For whatever reason trying to link to the sections results in a broken link, but updated

Reviewer 2: Hmmm. If it is busted, sometimes that happens if the title section has a space on the end. 

```
== Refresh <space here>
```

|500
|There was a FusionAuth internal error. A stack trace is provided and logged in the FusionAuth log files.
|===

> 

Reviewer 2: We should document the response cookies like we do here: https://fusionauth.io/docs/v1/tech/apis/login#authenticate-a-user

==== Response Body

include::docs/v1/tech/shared/_user-info-response.adoc[]

== Logout

> This API will start a logout. The cookies specific to this session will be removed and if an SSO session was started it will be ended.

Reviewer 2: ```suggestion
This API will start a logout. The cookies set on <<Login>> will be removed. If an SSO session was started, it will be ended.
```

> To use this API correctly the browser should be redirected to this route and it will respond with a `302` redirect status code if successful.

Reviewer 2: ```suggestion
To use this API, redirect the browser to this route. This is not meant to be called by non-browser clients.
```

No need to mention the 'if successful' since that is handled by the response table.

=== Request

[.api-authentication]
link:/docs/v1/tech/apis/authentication#no-authentication-required[icon:unlock[type=fas]] Start the logout flow

[.endpoint]
.URI
--
[method]#GET# [uri]#/app/logout/{clientId}``?redirect_uri=\{redirectUri\}``#
--

==== Request Parameters

[.api]
[field]#clientId# [type]#[UUID]# [required]#Required#::
The top-level client Id for your application.

[field]#redirect_uri# [type]#[String]# [optional]#Optional#::

> The URI that the browser will be redirected to at the end of the logout flow. If not provided this will be the `Logout URL` configured for the application. Other values must be defined in the `Authorized Redirect URLs` for the application.

Reviewer 2: ```suggestion
The URL encoded URL that the browser will be redirected to at the end of the logout flow. If not provided, this will be the [field]#Logout URL# configured for the Application. If another value is used, it must be defined in the [field]#Authorized Redirect URLs# for the Application.
```

.Example Request URL
[source,url]
----
https://auth.yourdomain.io/app/logout/297ca84b-69a9-4508-8649-97644e1d0b3d?redirect_uri=https%3A%2F%2Fapp.yourdomain.io%2
----

=== Response

> Successful invocations of this route will return a  `302` redirect to `/oauth2/logout`. Other status codes indicate an error. After logout the browser should be redirected to the defined `redirect_uri`.

Reviewer 2: ```suggestion
Successful invocations of this route will return a  `302` redirect to `/oauth2/logout`. Other status codes indicate an error. After logout the browser is redirected to the defined `redirect_uri`.
```

[cols="1a,9a"]
.Response Codes
|===
|Code |Description

|200
|This is likely an error case. The route will serve up an error page with html and details on what went wrong.

|302

> |A successful request will redirect the user to `/oauth2/authorize` to log in.

Reviewer 2: I don't understand this. Aren't we 302 redirecting to the logout url?
